### DISCUSSION TITLE
Projeto com dois tipos de usuários, faço 2 apps ou apenas uma com menu dinamico?

### DISCUSSION THREAD

**Discussion by wilsonneto-swe** [frontendbr/forum#1563] [Discussão]
Galera,

minha app terá dois tipos de usuários, o organizador que posta eventos, e as equipes que veem estes eventos e fazem notas e tarefas dali descritas.

Enfim, a parte do organizador/admin terá várias telas de crud e etc.

Já a equipe é mais simples, apenas ver e fazer certas notas, e dar tarefas como feitas.

Nestes casos como vocês fariam? Uma app (web) para cada caso ou uma maneira de gerar as rotas, proteger e mostrar menus específicos?

Como indicam fazer?

Meu medo é fazer a msm app e o usuário conseguir burlar editando algum dado javascript no browser... Por isso pensei em duas apps.

Abs.

**Comment by eduardosoareszup** [comment]
Pelo o que você explicou parece ser bem diferente as aplicações, talvez duas App realmente façam sentido. Mas compensa colocar na balança alguns pontos, por exemplo;

- O quanto de reaproveitamento de código você teria caso fosse só um?
- Aumentaria ou diminuiria a complexidade de evolução/manutenção do código apena um App?
- Quantos desenvolvedores estarão envolvidos no projeto? (Dependendo da quantidade dá para separar em duas squads separadas talvez)

Esses foram os pontos que pensei, mas o ideal é fazer um brainstorm com a equipe para ver o posicionamento de cada um, pois vocês são os melhores para responder essa pergunta, visto que estão ambientados com o projeto e sabem melhor do que nós as particularidades dele.

Espero ter ajudado e sucesso nesse projeto :rocket:

**Comment by ihenrique** [comment]
Cara sem o escopo é complicado, mais eu imaginei esse projeto assim, você tem um lado administrativo que gerencia e inclui eventos e outro lado que interage com esses eventos. Sendo assim faz todo o sentido separar, mais ou menos como o Wordpress faz, vc tem um adm isolado que só serve pra gestão de conteúdo e o front separado só pro usuário final com pouca ou quase nenhuma interação do usuário logado.

Só toma cuidado na organização do projeto, se for compartilhar componentes entre as duas partes, invista o tempo na documentação e no styleguide.

**Comment by eliseumds** [comment]
Eu separaria. Eventualmente as aplicações podem crescer e você vai ter que separá-las de qualquer forma. Hoje em dia é mais fácil compartilhar código entre diferentes apps usando um monorepo com Yarn Workspaces e Lerna, por exemplo. Benefícios:

* Liberdade para escolher diferentes dependência para cada app
* Build pipelines podem ser diferentes. Um app pode usar Webpack e o outro pode usar Parcel, por exemplo (não me parece ideal, mas é uma possibilidade)
* Fazer deploy de um app não afeta o outro
* Segurança, como você já citou. Não queremos expor todo o código do painel de administração no bundle do app público

Eu faria um esforço para evitar duplicação de rotas (de páginas e de API endpoints). Onde trabalho, temos um mapa YAML de rotas na raíz do projeto que é usado tanto pelas aplicações backend (PHP e NodeJS) quanto frontend. A vantagem é que você sempre referencia as rotas por um alias, um nome que não vai mudar:

```yml
public.home:
  path: /

public.user.profile:
  path: /user/{userId}
```

```yml
admin.home:
  path: /

admin.user.search:
  path: /user/search
```

```yml
api.admin.user.search:
  path: /api/user/search
  methods: [GET]

api.admin.user.details:
  path: /api/user/{userId}/details
  methods: [GET]
```

Depois compilamos isso para uns arquivos JS/JSON que podemos consumir no frontend:

```js
import getPath from 'shared/route-utils/getPath';
import publicRoutes from './algumarquivo.json';

<Link to={getPath(publicRoutes, 'public.user.settings', { userId: 'abc123' })}>
  See John's profile
</Link>
```

```js
import adminApiRoutes from './umoutroarquivo.json';

// Uma API call
const endpoint = getPath(adminApiRoutes, 'api.admin.user.details', { userId: 'abc123' });

fetch(endpoint).then(res => console.log(res));
```

Desta forma, você pode mudar a path sem ter que alterar nenhum outro arquivo.

**Comment by wilsonneto-swe** [comment]
Opa, valeu galera, ótimas observações, e valeu @eliseumds pelas observações e pela dicas das rotas... realmente pensando aqui isso vai ajudar.

Obrigado galera!

**Comment by ninetails** [comment]
@eliseumds só um off aqui: bem massa essa abstração pra rotas 👍 

> Meu medo é fazer a msm app e o usuário conseguir burlar editando algum dado javascript no browser... Por isso pensei em duas apps.

Se é uma preocupação segurança entre ambas aplicações, também sugeriria como o @eliseumds citou, monorepo ajudaria a transitar implementações e [modularizar a aplicação](https://en.wikipedia.org/wiki/Unix_philosophy#Do_One_Thing_and_Do_It_Well).

**Comment by felipefialho** [comment]
Ia falar de usar monorepo também, como sugerido pelo @ninetails, fica perfeito nessa situação.

No demais, as respostas do @eliseumds tem me representado aqui, tenho dado um 👍 atrás do outro, hahaha.
